File: error-logger-bot.js
const TelegramBot = require('node-telegram-bot-api');
const express = require('express');
const cors = require('cors');
const moment = require('moment');
const UAParser = require('ua-parser-js');

// Конфигурация
const BOT_TOKEN = '8087171595:AAGcTv_TiNAY-Mv8CSyaIwk2tzYnnEM4Dsk';
const ADMIN_ID = '7752180805'; // Только админ получает сообщения
const PORT = process.env.PORT || 3333;

// Инициализация бота
const bot = new TelegramBot(BOT_TOKEN, { polling: true });
const app = express();

// Middleware
app.use(cors());
app.use(express.json({ limit: '10mb' }));
app.use(express.urlencoded({ extended: true }));

console.log('🚀 Savolia Error Logger Bot запущен');
console.log(`👤 Админ ID: ${ADMIN_ID}`);
console.log(`🌐 Сервер запущен на порту: ${PORT}`);

// Класс для обработки ошибок
class ErrorLogger {
  constructor() {
    this.platforms = {
      'bot.ts': '🤖 Telegram Bot',
      'savolia-frontend': '📱 Mini App (Telegram)',
      'savolia-web': '🌐 Web Site'
    };
  }

  // Форматирование сообщения об ошибке
  formatErrorMessage(errorData) {
    const {
      platform,
      userPhone,
      device,
      osVersion,
      errorType,
      errorMessage,
      stackTrace,
      timestamp,
      url,
      userId,
      userAgent,
      additionalData
    } = errorData;

    const platformIcon = this.platforms[platform] || '❓ Unknown Platform';
    const formattedTime = moment(timestamp).format('DD.MM.YYYY HH:mm:ss');
    
    let message = `🚨 <b>ОШИБКА В SAVOLIA</b>\n\n`;
    message += `📍 <b>Платформа:</b> ${platformIcon}\n`;
    message += `📱 <b>Номер пользователя:</b> <code>${userPhone || 'Не указан'}</code>\n`;
    
    if (userId) {
      message += `🆔 <b>User ID:</b> <code>${userId}</code>\n`;
    }
    
    message += `💻 <b>Устройство:</b> ${device || 'Неизвестно'}\n`;
    message += `⚙️ <b>ОС:</b> ${osVersion || 'Неизвестно'}\n`;
    message += `❌ <b>Тип ошибки:</b> ${errorType || 'Общая ошибка'}\n`;
    message += `🕐 <b>Время:</b> ${formattedTime}\n\n`;
    
    if (url) {
      message += `🔗 <b>URL:</b> <code>${url}</code>\n\n`;
    }
    
    message += `📝 <b>Описание ошибки:</b>\n<code>${errorMessage}</code>\n\n`;
    
    if (stackTrace) {
      // Обрезаем слишком длинный stack trace
      const shortStack = stackTrace.length > 1000 
        ? stackTrace.substring(0, 1000) + '...\n[Обрезано]' 
        : stackTrace;
      message += `🔍 <b>Stack Trace:</b>\n<pre>${shortStack}</pre>\n\n`;
    }
    
    if (userAgent) {
      message += `🌐 <b>User Agent:</b>\n<code>${userAgent}</code>\n\n`;
    }
    
    if (additionalData && Object.keys(additionalData).length > 0) {
      message += `📊 <b>Дополнительные данные:</b>\n<pre>${JSON.stringify(additionalData, null, 2)}</pre>`;
    }
    
    return message;
  }

  // Парсинг User Agent для получения информации об устройстве
  parseUserAgent(userAgent) {
    if (!userAgent) return { device: 'Неизвестно', osVersion: 'Неизвестно' };
    
    const parser = new UAParser(userAgent);
    const result = parser.getResult();
    
    const device = `${result.device.vendor || ''} ${result.device.model || result.browser.name || 'Unknown'}`.trim();
    const osVersion = `${result.os.name || 'Unknown'} ${result.os.version || ''}`.trim();
    
    return { device, osVersion };
  }

  // Отправка ошибки админу
  async sendErrorToAdmin(errorData) {
    try {
      const message = this.formatErrorMessage(errorData);
      
      await bot.sendMessage(ADMIN_ID, message, {
        parse_mode: 'HTML',
        disable_web_page_preview: true
      });
      
      console.log(`✅ Ошибка отправлена админу: ${errorData.platform} - ${errorData.errorType}`);
      return true;
    } catch (error) {
      console.error('❌ Ошибка при отправке сообщения админу:', error);
      return false;
    }
  }
}

const errorLogger = new ErrorLogger();

// API эндпоинт для получения ошибок
app.post('/api/log-error', async (req, res) => {
  try {
    const {
      platform,
      userPhone,
      userId,
      errorType,
      errorMessage,
      stackTrace,
      url,
      userAgent,
      additionalData
    } = req.body;

    // Валидация обязательных полей
    if (!platform || !errorMessage) {
      return res.status(400).json({
        success: false,
        error: 'Обязательные поля: platform, errorMessage'
      });
    }

    // Парсинг User Agent для получения информации об устройстве
    const deviceInfo = errorLogger.parseUserAgent(userAgent);

    // Формирование данных ошибки
    const errorData = {
      platform,
      userPhone: userPhone || 'Не указан',
      userId,
      device: deviceInfo.device,
      osVersion: deviceInfo.osVersion,
      errorType: errorType || 'Runtime Error',
      errorMessage,
      stackTrace,
      timestamp: new Date().toISOString(),
      url,
      userAgent,
      additionalData
    };

    // Отправка ошибки админу
    const sent = await errorLogger.sendErrorToAdmin(errorData);

    if (sent) {
      res.json({
        success: true,
        message: 'Ошибка успешно зарегистрирована'
      });
    } else {
      res.status(500).json({
        success: false,
        error: 'Не удалось отправить ошибку'
      });
    }

  } catch (error) {
    console.error('❌ Ошибка в API:', error);
    res.status(500).json({
      success: false,
      error: 'Внутренняя ошибка сервера'
    });
  }
});

// Health check эндпоинт
app.get('/health', (req, res) => {
  res.json({
    status: 'OK',
    timestamp: new Date().toISOString(),
    service: 'Savolia Error Logger Bot'
  });
});

// Обработка сообщений от пользователей
bot.on('message', async (msg) => {
  const chatId = msg.chat.id;
  const userId = msg.from.id.toString();

  // Только админ может взаимодействовать с ботом
  if (userId !== ADMIN_ID) {
    // Не отвечаем ничем другим пользователям
    console.log(`🚫 Неавторизованный доступ от пользователя: ${userId}`);
    return;
  }

  // Команды для админа
  if (msg.text === '/start') {
    await bot.sendMessage(chatId, 
      '🚀 <b>Savolia Error Logger Bot</b>\n\n' +
      '🔍 Этот бот отслеживает ошибки во всех платформах Savolia:\n' +
      '• 🤖 Telegram Bot (bot.ts)\n' +
      '• 📱 Mini App (savolia-frontend)\n' +
      '• 🌐 Web Site (savolia-web)\n\n' +
      '📊 <b>Доступные команды:</b>\n' +
      '/status - Статус бота\n' +
      '/help - Справка\n\n' +
      '🔔 Все ошибки будут автоматически пересылаться сюда.',
      { parse_mode: 'HTML' }
    );
  } else if (msg.text === '/status') {
    await bot.sendMessage(chatId,
      '✅ <b>Статус системы:</b>\n\n' +
      `🤖 Бот: Активен\n` +
      `🌐 API: Работает на порту ${PORT}\n` +
      `🕐 Время работы: ${moment().format('DD.MM.YYYY HH:mm:ss')}\n` +
      `📊 Платформы: ${Object.keys(errorLogger.platforms).length}`,
      { parse_mode: 'HTML' }
    );
  } else if (msg.text === '/help') {
    await bot.sendMessage(chatId,
      '📖 <b>Справка по Error Logger Bot</b>\n\n' +
      '<b>Формат ошибки:</b>\n' +
      '• Платформа (bot.ts/frontend/web)\n' +
      '• Номер пользователя\n' +
      '• Устройство и ОС\n' +
      '• Тип и описание ошибки\n' +
      '• Stack trace\n' +
      '• Время возникновения\n\n' +
      '<b>API Endpoint:</b>\n' +
      '<code>POST /api/log-error</code>\n\n' +
      'Все ошибки автоматически форматируются и отправляются в этот чат.',
      { parse_mode: 'HTML' }
    );
  }
});

// Обработка ошибок бота
bot.on('error', (error) => {
  console.error('❌ Ошибка Telegram бота:', error);
});

// Graceful shutdown
process.on('SIGINT', () => {
  console.log('\n🛑 Завершение работы Error Logger Bot...');
  bot.stopPolling();
  process.exit(0);
});

// Запуск сервера
app.listen(PORT, () => {
  console.log(`🌐 API сервер запущен на http://localhost:${PORT}`);
  console.log(`📡 Эндпоинт для ошибок: POST /api/log-error`);
});

File: error_logger_bot.py
import asyncio
import json
import logging
import os
from datetime import datetime
from typing import Dict, Any, Optional

import requests
from flask import Flask, request, jsonify
from flask_cors import CORS
from telegram import Bot
from telegram.ext import Application, CommandHandler, MessageHandler, filters
from user_agents import parse as parse_user_agent
import threading

# Конфигурация
BOT_TOKEN = '8087171595:AAGcTv_TiNAY-Mv8CSyaIwk2tzYnnEM4Dsk'
ADMIN_ID = '7752180805'  # Только админ получает сообщения
PORT = int(os.environ.get('PORT', 3333))

# Настройка логирования
logging.basicConfig(
    format='%(asctime)s - %(name)s - %(levelname)s - %(message)s',
    level=logging.INFO
)
logger = logging.getLogger(__name__)

# Flask приложение
app = Flask(__name__)
CORS(app)

class ErrorLogger:
    def __init__(self):
        self.platforms = {
            'bot.ts': '🤖 Telegram Bot',
            'savolia-frontend': '📱 Mini App (Telegram)',
            'savolia-web': '🌐 Web Site'
        }
        self.bot = None
        
    def set_bot(self, bot: Bot):
        """Устанавливает экземпляр бота"""
        self.bot = bot
    
    def format_error_message(self, error_data: Dict[str, Any]) -> str:
        """Форматирование сообщения об ошибке"""
        platform = error_data.get('platform', 'unknown')
        user_phone = error_data.get('userPhone', 'Не указан')
        device = error_data.get('device', 'Неизвестно')
        os_version = error_data.get('osVersion', 'Неизвестно')
        error_type = error_data.get('errorType', 'Общая ошибка')
        error_message = error_data.get('errorMessage', 'Нет описания')
        stack_trace = error_data.get('stackTrace')
        timestamp = error_data.get('timestamp')
        url = error_data.get('url')
        user_id = error_data.get('userId')
        user_agent = error_data.get('userAgent')
        additional_data = error_data.get('additionalData')
        
        platform_icon = self.platforms.get(platform, '❓ Unknown Platform')
        
        # Форматирование времени
        if timestamp:
            try:
                dt = datetime.fromisoformat(timestamp.replace('Z', '+00:00'))
                formatted_time = dt.strftime('%d.%m.%Y %H:%M:%S')
            except:
                formatted_time = timestamp
        else:
            formatted_time = datetime.now().strftime('%d.%m.%Y %H:%M:%S')
        
        message = f"🚨 <b>ОШИБКА В SAVOLIA</b>\n\n"
        message += f"📍 <b>Платформа:</b> {platform_icon}\n"
        message += f"📱 <b>Номер пользователя:</b> <code>{user_phone}</code>\n"
        
        if user_id:
            message += f"🆔 <b>User ID:</b> <code>{user_id}</code>\n"
            
        message += f"💻 <b>Устройство:</b> {device}\n"
        message += f"⚙️ <b>ОС:</b> {os_version}\n"
        message += f"❌ <b>Тип ошибки:</b> {error_type}\n"
        message += f"🕐 <b>Время:</b> {formatted_time}\n\n"
        
        if url:
            message += f"🔗 <b>URL:</b> <code>{url}</code>\n\n"
            
        message += f"📝 <b>Описание ошибки:</b>\n<code>{error_message}</code>\n\n"
        
        if stack_trace:
            # Обрезаем слишком длинный stack trace
            short_stack = stack_trace[:1000] + '...\n[Обрезано]' if len(stack_trace) > 1000 else stack_trace
            message += f"🔍 <b>Stack Trace:</b>\n<pre>{short_stack}</pre>\n\n"
            
        if user_agent:
            message += f"🌐 <b>User Agent:</b>\n<code>{user_agent}</code>\n\n"
            
        if additional_data and isinstance(additional_data, dict) and additional_data:
            formatted_data = json.dumps(additional_data, ensure_ascii=False, indent=2)
            message += f"📊 <b>Дополнительные данные:</b>\n<pre>{formatted_data}</pre>"
            
        return message
    
    def parse_user_agent(self, user_agent: Optional[str]) -> Dict[str, str]:
        """Парсинг User Agent для получения информации об устройстве"""
        if not user_agent:
            return {'device': 'Неизвестно', 'osVersion': 'Неизвестно'}
            
        try:
            parsed = parse_user_agent(user_agent)
            
            # Формируем название устройства
            device_parts = []
            if parsed.device.brand:
                device_parts.append(parsed.device.brand)
            if parsed.device.model:
                device_parts.append(parsed.device.model)
            if not device_parts and parsed.browser.family:
                device_parts.append(parsed.browser.family)
                
            device = ' '.join(device_parts) if device_parts else 'Неизвестно'
            
            # Формируем версию ОС
            os_parts = []
            if parsed.os.family:
                os_parts.append(parsed.os.family)
            if parsed.os.version_string:
                os_parts.append(parsed.os.version_string)
                
            os_version = ' '.join(os_parts) if os_parts else 'Неизвестно'
            
            return {'device': device, 'osVersion': os_version}
            
        except Exception as e:
            logger.error(f"Ошибка парсинга User Agent: {e}")
            return {'device': 'Неизвестно', 'osVersion': 'Неизвестно'}
    
    async def send_error_to_admin(self, error_data: Dict[str, Any]) -> bool:
        """Отправка ошибки админу"""
        if not self.bot:
            logger.error("Бот не инициализирован")
            return False
            
        try:
            message = self.format_error_message(error_data)
            
            await self.bot.send_message(
                chat_id=ADMIN_ID,
                text=message,
                parse_mode='HTML',
                disable_web_page_preview=True
            )
            
            logger.info(f"✅ Ошибка отправлена админу: {error_data.get('platform')} - {error_data.get('errorType')}")
            return True
            
        except Exception as e:
            logger.error(f"❌ Ошибка при отправке сообщения админу: {e}")
            return False

# Создаем экземпляр логгера
error_logger = ErrorLogger()

# Telegram бот handlers
async def start_command(update, context):
    """Обработчик команды /start"""
    user_id = str(update.effective_user.id)
    
    if user_id != ADMIN_ID:
        logger.warning(f"🚫 Неавторизованный доступ от пользователя: {user_id}")
        return
    
    message = (
        "🚀 <b>Savolia Error Logger Bot</b>\n\n"
        "🔍 Этот бот отслеживает ошибки во всех платформах Savolia:\n"
        "• 🤖 Telegram Bot (bot.ts)\n"
        "• 📱 Mini App (savolia-frontend)\n"
        "• 🌐 Web Site (savolia-web)\n\n"
        "📊 <b>Доступные команды:</b>\n"
        "/status - Статус бота\n"
        "/help - Справка\n\n"
        "🔔 Все ошибки будут автоматически пересылаться сюда."
    )
    
    await update.message.reply_text(message, parse_mode='HTML')

async def status_command(update, context):
    """Обработчик команды /status"""
    user_id = str(update.effective_user.id)
    
    if user_id != ADMIN_ID:
        return
    
    current_time = datetime.now().strftime('%d.%m.%Y %H:%M:%S')
    platforms_count = len(error_logger.platforms)
    
    message = (
        "✅ <b>Статус системы:</b>\n\n"
        f"🤖 Бот: Активен\n"
        f"🌐 API: Работает на порту {PORT}\n"
        f"🕐 Время работы: {current_time}\n"
        f"📊 Платформы: {platforms_count}"
    )
    
    await update.message.reply_text(message, parse_mode='HTML')

async def help_command(update, context):
    """Обработчик команды /help"""
    user_id = str(update.effective_user.id)
    
    if user_id != ADMIN_ID:
        return
    
    message = (
        "📖 <b>Справка по Error Logger Bot</b>\n\n"
        "<b>Формат ошибки:</b>\n"
        "• Платформа (bot.ts/frontend/web)\n"
        "• Номер пользователя\n"
        "• Устройство и ОС\n"
        "• Тип и описание ошибки\n"
        "• Stack trace\n"
        "• Время возникновения\n\n"
        "<b>API Endpoint:</b>\n"
        "<code>POST /api/log-error</code>\n\n"
        "Все ошибки автоматически форматируются и отправляются в этот чат."
    )
    
    await update.message.reply_text(message, parse_mode='HTML')

async def handle_message(update, context):
    """Обработчик обычных сообщений"""
    user_id = str(update.effective_user.id)
    
    if user_id != ADMIN_ID:
        # Не отвечаем ничем другим пользователям
        logger.warning(f"🚫 Неавторизованное сообщение от пользователя: {user_id}")
        return

# Flask API endpoints
@app.route('/api/log-error', methods=['POST'])
def log_error():
    """API эндпоинт для получения ошибок"""
    try:
        data = request.get_json()
        
        if not data:
            return jsonify({
                'success': False,
                'error': 'Нет данных в запросе'
            }), 400
        
        # Валидация обязательных полей
        platform = data.get('platform')
        error_message = data.get('errorMessage')
        
        if not platform or not error_message:
            return jsonify({
                'success': False,
                'error': 'Обязательные поля: platform, errorMessage'
            }), 400
        
        # Парсинг User Agent
        user_agent = data.get('userAgent')
        device_info = error_logger.parse_user_agent(user_agent)
        
        # Формирование данных ошибки
        error_data = {
            'platform': platform,
            'userPhone': data.get('userPhone', 'Не указан'),
            'userId': data.get('userId'),
            'device': device_info['device'],
            'osVersion': device_info['osVersion'],
            'errorType': data.get('errorType', 'Runtime Error'),
            'errorMessage': error_message,
            'stackTrace': data.get('stackTrace'),
            'timestamp': data.get('timestamp', datetime.now().isoformat()),
            'url': data.get('url'),
            'userAgent': user_agent,
            'additionalData': data.get('additionalData')
        }
        
        # Отправка ошибки админу асинхронно
        loop = asyncio.new_event_loop()
        asyncio.set_event_loop(loop)
        sent = loop.run_until_complete(error_logger.send_error_to_admin(error_data))
        loop.close()
        
        if sent:
            return jsonify({
                'success': True,
                'message': 'Ошибка успешно зарегистрирована'
            })
        else:
            return jsonify({
                'success': False,
                'error': 'Не удалось отправить ошибку'
            }), 500
            
    except Exception as e:
        logger.error(f"❌ Ошибка в API: {e}")
        return jsonify({
            'success': False,
            'error': 'Внутренняя ошибка сервера'
        }), 500

@app.route('/health', methods=['GET'])
def health_check():
    """Health check эндпоинт"""
    return jsonify({
        'status': 'OK',
        'timestamp': datetime.now().isoformat(),
        'service': 'Savolia Error Logger Bot'
    })

def run_bot():
    """Запуск Telegram бота"""
    async def main():
        # Создание приложения
        application = Application.builder().token(BOT_TOKEN).build()
        
        # Устанавливаем бота в логгер
        error_logger.set_bot(application.bot)
        
        # Добавление обработчиков
        application.add_handler(CommandHandler("start", start_command))
        application.add_handler(CommandHandler("status", status_command))
        application.add_handler(CommandHandler("help", help_command))
        application.add_handler(MessageHandler(filters.TEXT & ~filters.COMMAND, handle_message))
        
        # Запуск бота
        logger.info("🚀 Savolia Error Logger Bot запущен")
        logger.info(f"👤 Админ ID: {ADMIN_ID}")
        
        await application.run_polling()
    
    # Запуск в отдельном потоке
    asyncio.run(main())

if __name__ == '__main__':
    # Запуск бота в отдельном потоке
    bot_thread = threading.Thread(target=run_bot, daemon=True)
    bot_thread.start()
    
    # Запуск Flask сервера
    logger.info(f"🌐 API сервер запущен на http://localhost:{PORT}")
    logger.info(f"📡 Эндпоинт для ошибок: POST /api/log-error")
    
    app.run(host='0.0.0.0', port=PORT, debug=False)